File: 2048-next-greater-numerically-balanced-number.py
from itertools import permutations, combinations
from bisect import bisect_right

class Solution:
    def nextBeautifulNumber(self, n: int) -> int:
        '''
        We will first generate a list of beautiful numbers. It should be a small list. After that we just go through each number and return the first number greater than n
        '''
        def gen_beautiful_numbers() -> list[int]:
            bases = [str(i) * i for i in range(1, 7)]

            beautiful_numbers = set()
            limit = 10**7

            for size in range(1, len(bases) + 1):
                for combo in combinations(bases, size):
                    combined = ''.join(combo)
                    if len(combined) > 7:
                        continue
                    for p in set(permutations(combined)):
                        temp = int("".join(p))
                        if temp < limit:
                            beautiful_numbers.add(temp)

            for base in bases:
                beautiful_numbers.add(int(base))

            return sorted(beautiful_numbers)
            
        numbers = gen_beautiful_numbers()
        idx = bisect_right(numbers, n)
        return numbers[idx]


sol = Solution()
print(f'output: {sol.nextBeautifulNumber(1)}, expected: 22')
print(f'output: {sol.nextBeautifulNumber(1000)}, expected: 1333')
print(f'output: {sol.nextBeautifulNumber(3000)}, expected: 3133')
print(f'output: {sol.nextBeautifulNumber(59866)}, expected: 122333')
print(f'output: {sol.nextBeautifulNumber(748601)}, expected: 1224444')

File: 2048-next-greater-numerically-balanced-number.java
import java.util.*;

class NextGreaterNumericallyBalancedNumber {
  
    public int nextBeautifulNumber(int n) {
        List<Integer> numbers = genBeautifulNumbers();
        for (Integer integer : numbers) {
            if (integer > n) {
                return integer;
            }
        }
        return -1;
    }

    private List<Integer> genBeautifulNumbers() {
        List<String> bases = new ArrayList<>();
        for (int i = 1; i <= 6; i++) {
            bases.add(String.valueOf(i).repeat(i));
        }
        Set<Integer> beautifulNumbers = new HashSet<>();
        int limit = 10_000_000;

        for (int size = 1; size <= bases.size(); size++) {
            for (List<String> combo : combinations(bases, size)) {
                String combined = String.join("", combo);
                
                // Early skip if combined string is too long
                if (combined.length() > 7) {
                    continue;
                }
                
                // Generate all permutations
                for (String perm : permutations(combined)) {
                    int num = Integer.parseInt(perm);
                    if (num < limit) {
                        beautifulNumbers.add(num);
                    }
                }
            }
        }
        
        List<Integer> result = new ArrayList<>(beautifulNumbers);
        for (String base : bases) {
            result.add(Integer.valueOf(base));
        }
        Collections.sort(result);
        return result;
    }

    private static <T> List<List<T>> combinations(List<T> list, int k) {
        List<List<T>> result = new ArrayList<>();
        combinationsHelper(list, k, 0, new ArrayList<>(), result);
        return result;
    }
    
    private static <T> void combinationsHelper(List<T> list, int k, int start, 
                                                List<T> current, List<List<T>> result) {
        if (current.size() == k) {
            result.add(new ArrayList<>(current));
            return;
        }
        
        for (int i = start; i < list.size(); i++) {
            current.add(list.get(i));
            combinationsHelper(list, k, i + 1, current, result);
            current.remove(current.size() - 1);
        }
    }
    
    // Generate all permutations of a string
    private static Set<String> permutations(String str) {
        Set<String> result = new HashSet<>();
        permutationsHelper(str.toCharArray(), 0, result);
        return result;
    }
    
    private static void permutationsHelper(char[] chars, int index, Set<String> result) {
        if (index == chars.length - 1) {
            result.add(new String(chars));
            return;
        }
        
        for (int i = index; i < chars.length; i++) {
            swap(chars, i, index);
            permutationsHelper(chars, index + 1, result);
            swap(chars, i, index); // backtrack
        }
    }
    
    private static void swap(char[] chars, int i, int j) {
        char temp = chars[i];
        chars[i] = chars[j];
        chars[j] = temp;
    }

    public static void main(String[] args) {
        NextGreaterNumericallyBalancedNumber sol = new NextGreaterNumericallyBalancedNumber();
        System.out.println("Output: " + sol.nextBeautifulNumber(1) + ", expected: 22");
        System.out.println("Output: " + sol.nextBeautifulNumber(1000) + ", expected: 1333");
        System.out.println("Output: " + sol.nextBeautifulNumber(3000) + ", expected: 3133");
        System.out.println("Output: " + sol.nextBeautifulNumber(59866) + ", expected: 122333");
        System.out.println("Output: " + sol.nextBeautifulNumber(748601) + ", expected: 1224444");
    }
}
